### ISSUE TITLE
Incremental Tile updates

### ISSUE THREAD

**Issue by zstadler** [IsraelHikingMap/Map#108]
This ticket is intended to be a place to discuss the possibility of incrementally updating the Israel Hiking and Biking maps.

### Motivation
Currently, the update process for both maps takes 32-36 hours. As a result, the process is run every other day. It therefore can take 2-3 days from a map edit until the change is reflected in the maps.

### Efficiency mechanisms used in the current process
- The "fingerprint" mechanism in Maperitive writes a checksum for each tile. When creating a new map, the checksum of the internally-rendered tile is compared with the checksum of the tile on disk. The tile is written to the disk only if the checksums differ.
- [GenIsraelHikingTiles.py](/IsraelHikingMap/Map/blob/17ed10cc24093595946c9f747d07725836064fa2/Scripts/Maperipy/GenIsraelHikingTiles.py) limits the map generation to the Israel polygon in zoom levels 12 and above. The Maperipy tile generation API is used to reduce rendering of chunks of neighboring tiles (using ```generation_filter```) and individual tile writing to disk (using ```save_filter```). 

### Disadvantages of the current process
- The internal rendering is performed for all areas within the Israel polygon at all zoom levels.
- The internal rendering is inconsistent and about 10% of the tiles receive different checksums for the when created from the same map sources.

### Proposed Approach
On most days, the OSM data changes are localized to a few areas. For replication purposes, OSM published the [diffs to its database](https://wiki.openstreetmap.org/wiki/Planet.osm/diffs) for every minute, hour and day in the form of a [OsmChange](https://wiki.openstreetmap.org/wiki/OsmChange) files.

An Incremental Tile Generation process is proposed to leverage this diff information to reduce the rendering performed by Maperitive to the subset of the tiles that can be affected by a given Osm Change file.

### Benchmarks
- Complete map creation process with no actual rendering takes about 21 minutes.
- Complete map creation process with no tile saving takes about the same time as the bi-daily process

Conclusion: There is a significant potential for saving  time if rendering is limited to tiles surrounding the changes in the OSM data.

### Open issues
- How to identify the affected tiles? 
While the location of nodes is included in the Osm Change file, the location of ways and relations is not included.
- How to update the pbf file with the Osm Change file?
  Take care to avoid an explosion in pbf file size.
- How to avoid running the map update when no changes were made in the Israel polygon?

**Comment by HarelM** [comment]
Server code is using osmsharp which uses porbuf.net which is the database framework, I'm sure it won't be hard to do it add part of the updater code which already downloads and parses the pbf file.

My main concern is inconsistency in maperitive partial tiles generation.

**Comment by HarelM** [comment]
This should cover merges of files:
http://stackoverflow.com/questions/3453552/protobuf-net-merge-collection-to-existing-file
I also think the the incremental file can be opened using OSM sharp and this might help in understanding which tiles might need an update - current image creation on server knows which tile belongs to which coordinate so knowing which tiles to recreate might not prove hard.

I also think that we might want to run it once a week or once a month from a clean version so that it wouldn't be patches all the time...?

**Comment by valleyofdawn** [comment]
Regardless of incrementality,  can't several copies of the renderer run in
parallel on different areas of the land?

**Comment by HarelM** [comment]
It might burden the file system on the server.
I think we should consider moving to Mapnik. It's open source highly maintained and cross platform. It improved a lot since I visited the decision to use maperitive over it.

**Comment by zstadler** [comment]
Issue #93 should cover the discussion about Maperitive alternatives, including MapsForge and Mapnik.

**Comment by HarelM** [comment]
Not sure we can split the discussion. Incremental tiles update requires that the process of creating tiles will be definite - meaning if I run the process of creating a map on a tile that was not updated it should be created exactly the same (Recall the chess board issue with Trailze). Correct me if I'm wrong, but as far as I know Maperitive has an issue with this.
A visible example of it is the overlay hiking trails (I hope it's a valid example) - when you show them on the IHM map you should not see any difference but the fact is that the trails get "bold" since it's not exactly one on top of the other.
So, IMO, if we want to accomplish incremental tiles updates we might need to migrate to another map generation alternative.

**Comment by HarelM** [comment]
@valleyofdawn - see issue #42 for parallel run. Note that this might improve the time it takes to create the entire map the fact is that you don't need to create the entire map every day or so, you only need to create the relevant tiles that were changed, the difference in time is a few magnitudes no matter how fast you can build the entire map...

**Comment by zstadler** [comment]
@HarelM 
The goal of this issue is to design, and hopefully implement, a Maperitive-based incremental solution.
Feel free to open separate issue/s for the alternative approaches so we can compare feasibility, complexity, and effort.

The chessboard issue was a result of the tiles not kept up-to-date on the secondary server. For this reason, we no longer use it. It is unrelated to Maperitive.

In the past, the rules for the map have evolved and the rules for the overlay. Commit 8c1378896f5c2a1edfdc330396e613c094d576aa aligned them. AFAIK, even the gaps in dashed lines are in the same locations. Note that overlaying the trails on to of the hiking map creates sharper lines because the anti-aliasing effect is halved by having two layers. The effect is easier to see in zoom 19 or higher.

@valleyofdawn 
The goal is to increase the efficiency of the map update by computing less. The efficiency improvement by updating only potentially-affected tiles can be several orders of magnitude.

**Comment by HarelM** [comment]
OK, I don't mind being wrong about this. Assuming I'll take care of the pbf download and merge as part of the routing update. And assuming I can understand which tiles need update from the delta pbf file, how would you like this updater to communicate the information with Maperitive? a file written to disk? command-line arguments?

**Comment by zstadler** [comment]
The information about which tiles need to be updated should be available while Maperitive's ```generate-tiles``` command is executing. Therefore, the preferred approach would be to implement the entire mechanism as an improvement of the [GenIsraelHikingTiles.py](/IsraelHikingMap/Map/blob/17ed10cc24093595946c9f747d07725836064fa2/Scripts/Maperipy/GenIsraelHikingTiles.py) Maperipy script. This script was also used for the above benchmark, so we know it can effectively control the rendering.

In order to minimize re-implementation, external libraries or programs, such as [Osmconvert](https://wiki.openstreetmap.org/wiki/Osmconvert), would be used to perform certain operations, like extracting the change info relevant to the map outline polygon and merging these changes into a new pbf file.

**Comment by HarelM** [comment]
Using third party library is great, the only question is if it satisfies all our needs.
Also note that if we put it inside the GenIsraelHikingTiles.py we'll need to add the graphhopper update there as well in order to avoid un-synced routing and tiles.
Assuming you don't have issues with all the above feel free to start implementing.

**Comment by ATGardner** [comment]
I was looking into trying something like this a while ago. Wouldn't a change file from [geofabrik](http://download.geofabrik.de/asia/israel-and-palestine.html) be enough? I already got a list of tiles that needs updating, from the change files. What I missing something?

One problem I anticipated was that even though a node located in tile x,y,z can change, which will cause it to re-render, the label field (whichever it may be) may "spill" over to adjacent nodes, and it won't be clear if and when it happens. This might cause labels to be cut off between tiles (though I am not sure - maybe Maperitive already renders all labels inside the tile that contains the node? - that seems like a simple enough solution).

**Comment by zstadler** [comment]
I'm currently writing the Osm Change file analysis code. 
Can you share your code that analyzes the change file? 
Does it handle deletion of ways and relations?

As for labels appearing on adjacent tiles, the proposed approach is to add all adjacent tiles for the tiles that contain an actual change. This should be done independently at each zoom level.

**Comment by zstadler** [comment]
By the way, for our region, geofabrik produces daily change files. If processing time allows, I would like to run the update several times a day by combining a few hourly change files.

**Comment by ATGardner** [comment]
This is a small POC a friend of mine did, while helping me understand some basic python. It just reads all the nodes lat/lon, and prints a list of tiles at zoom 15, and a list of center coordinates of those tiles. It does not get way/relation info from OSM. I don't think it will be of much help, but here it is, regardless.

[read_coordinates.zip](https://github.com/IsraelHikingMap/Map/files/592706/read_coordinates.zip)

**Comment by zstadler** [comment]
Thanks for the code. As you said, it is a partial implementation.
I also think xml.etree.ElementTree does not work without expat on IronPython, including Maperipy.

**Comment by zstadler** [comment]
At first stage, the incremental update continues to use Geofabrik's daily updates. 
This stage will allow us to check the results of the incremental update, the consistency of the resulting map over time, and the performance of the system.
At this stage, there is also no need to change the update process for the site search and routing services.

[Osmupdate](https://wiki.openstreetmap.org/wiki/Osmupdate) downloads the required change data directly form a replication data server and creates an Osm Change file:
```
osmup.exe israel-and-palestine-latest.osm.pbf israel-and-palestine-update.osc ^
  --base-url=download.geofabrik.de/asia/israel-and-palestine
```

Osmconvert will then create the updated pbf file:
```
osconvert.exe israel-and-palestine-latest.osm.pbf israel-and-palestine-update.osc ^
  --o=israel-and-palestine-update.osm.pbf
```

The load of the 3 input files and the change analysis is done once for all maps.

The next stage would be to enter a continuous update process using minute diffs:
```
osmup.exe israel-and-palestine-latest.osm.pbf israel-and-palestine-update.osc ^
  --base-url=download.openstreetmap.fr/replication/asia/israel
```

**Comment by HarelM** [comment]
Once the data is loaded to Maperive you just apply the relevant rules set and generate the relevant tiles?
I'm asking this because this can open up opportunities to create more maps since the the update and tile generation can be very quick.
The first candidate is the tourists map from issue #8 
I wouldn't hurry to minutes, I think hours might be enough at this point.

**Comment by zstadler** [comment]
IMO, 
- Completing an update in a couple of hours is critical priority
- Reducing it to less than an hour is high priority
- Adding non-Hebrew maps is low priority

Once a mechanism for non-Hebrew maps is in place, the maps can be updated on a monthly basis with no significant impact on users or the server. Weekly updates would also be possible if we see an increase in traffic.

**Comment by zstadler** [comment]
Incremental tile generation is deployed with [download.geofabrik.de](http://download.geofabrik.de) updates.

The generation was completed in 4.5 hours instead of an average of 28 hours.
